### DISCUSSION TITLE
Error: It is not allowed to define inline "use server" annotated Server Actions in Client Components.

### DISCUSSION THREAD

**Discussion by dhairyathedev** [vercel/next.js#61623] [Help]
### Summary

Unable to use `use server` inside a client component.
<img width="957" alt="Error Screenshot" src="https://github.com/vercel/next.js/assets/65452005/18affbc0-96d0-49ff-a63d-e90cd730e17e">

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:27 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T8103
Binaries:
  Node: 20.5.1
  npm: 9.8.0
  Yarn: 1.22.21
  pnpm: N/A
Relevant Packages:
  next: 14.1.0
  eslint-config-next: 14.1.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

**Comment by icyJoseph** [comment]
Well it is pretty clear isn' it? Move it to its own file, and export it from there. The reason being that a server action is server code, which can be called from the client. However, the client bundle must not have access to that code to being with, in order to prevent leakage to the client.

Read more here:

- https://react.dev/reference/react/use-server#caveats

**Reply by dhairyathedev** [reply]
Thanks! I have fixed the error :)

**Reply by sojinsamuel** [reply]
Thnx for the explanation

**Reply by vedant-asati** [reply]
@dhairyathedev  how did you fix this? can you share the code... im having issues with implementing the signin function. If I am implementing in server comps.. im getting errs, as it needs client context.

![image](https://github.com/user-attachments/assets/92340237-6ea8-46cc-85ad-a55a3804c1af)

**Reply by aryaadinulfadlan** [reply]
> Well it is pretty clear isn' it? Move it to its own file, and export it from there. The reason being that a server action is server code, which can be called from the client. However, the client bundle must not have access to that code to being with, in order to prevent leakage to the client.
> 
> Read more here:
> 
> * https://react.dev/reference/react/use-server#caveats

Hi @icyJoseph ,

I did what you suggested, which was separating the function into a different file and adding the "use server" flag at the top of that file.

However, the same error happens again when I don’t put the "use server" flag at the top of the file but place it inside the function instead. What’s wrong? I’ve already moved this function into a separate file, but the error still occurs because the "use server" flag is inside the function.

Does the "use server" flag always have to be at the very top of the file?

**Comment by mfgabriel92** [comment]
Why is this way still in the official training if it doesn't work?

**Comment by jhourlad01** [comment]
Can anybody share the actual code to implement this? Thanks.

**Reply by icyJoseph** [reply]
A server action has to be defined in server code, and be consumed by a client module, not owned by it. Be it as an import or props.
